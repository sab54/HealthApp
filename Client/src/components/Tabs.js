// Client/src/components/Tabs.js
/**
 * Tabs.js
 * 
 * This file defines the `Tabs` component, which displays a row of tab items, each of which can 
 * be selected. The component supports animated tab indicators, customizable colors, and optional 
 * scrollability. It allows users to switch between different tabs, with the active tab being 
 * highlighted and an animated indicator moving to the selected tab. The tabs are designed to 
 * be responsive, with support for both fixed and scrollable layouts.
 * 
 * Features:
 * - Displays a list of tabs that can be selected.
 * - Animated indicator that moves smoothly between the selected tabs.
 * - Customizable theme for colors, including text, background, and indicator color.
 * - Support for scrollable tabs if there are too many to fit in a single row.
 * - Disabled state for tabs to prevent selection.
 * - Accessibility support, including roles and states for screen readers.
 * 
 * Props:
 * - `tabs`: An array of tab objects, where each object contains a `key`, `label`, and an optional `disabled` flag.
 * - `selectedTab`: The key of the currently selected tab.
 * - `onTabSelect`: Callback function that is triggered when a tab is selected.
 * - `theme`: The theme object for customizing colors like `primary`, `text`, `muted`, and `surface`.
 * - `scrollable`: Boolean flag to enable horizontal scrolling for the tabs (default is `false`).
 * - `indicatorColor`: The color of the animated indicator (defaults to `theme.primary`).
 * 
 * Dependencies:
 * - `react-native`
 * 
 * Author: Sunidhi Abhange
 */

import React, { useEffect, useRef, useState } from 'react';
import {
    View,
    TouchableOpacity,
    Text,
    StyleSheet,
    Animated,
    ScrollView,
} from 'react-native';

const Tabs = ({
    tabs,
    selectedTab,
    onTabSelect,
    theme = {},
    scrollable = false,
    indicatorColor = theme?.primary || '#4B7BE5',
}) => {
    const indicatorAnim = useRef(new Animated.Value(0)).current;
    const containerWidth = useRef(0);
    const [tabWidth, setTabWidth] = useState(0);

    const handleLayout = (e) => {
        const width = e.nativeEvent.layout.width;
        containerWidth.current = width;
        setTabWidth(width / tabs.length);
    };

    useEffect(() => {
        const index = tabs.findIndex((tab) => tab.key === selectedTab);
        if (index !== -1) {
            Animated.spring(indicatorAnim, {
                toValue: index,
                useNativeDriver: false,
                speed: 20,
                bounciness: 8,
            }).start();
        }
    }, [selectedTab, tabs]);

    const renderTabs = () =>
        tabs.map((tab, index) => {
            const isSelected = selectedTab === tab.key;
            const isDisabled = tab.disabled;

            return (
                <TouchableOpacity
                    key={tab.key}
                    style={styles.tabItem}
                    onPress={() => !isDisabled && onTabSelect(tab.key)}
                    activeOpacity={isDisabled ? 1 : 0.8}
                    accessibilityRole='button'
                    accessibilityState={{
                        selected: isSelected,
                        disabled: isDisabled,
                    }}
                >
                    <Text
                        style={[
                            styles.tabText,
                            {
                                color: isDisabled
                                    ? theme.muted || '#999'
                                    : isSelected
                                    ? theme.primary || '#4B7BE5'
                                    : theme.text || '#000',
                                fontWeight: isSelected ? '600' : '500',
                            },
                        ]}
                    >
                        {tab.label}
                    </Text>
                </TouchableOpacity>
            );
        });

    const translateX = indicatorAnim.interpolate({
        inputRange: [0, tabs.length - 1],
        outputRange: [0, tabWidth * (tabs.length - 1)],
    });

    return (
        <View style={styles.wrapper} onLayout={handleLayout}>
            <ScrollView
                horizontal={scrollable}
                contentContainerStyle={[
                    styles.tabRow,
                    {
                        backgroundColor: theme.surface || '#f5f5f5',
                        borderColor: theme.border || '#ddd',
                    },
                ]}
                showsHorizontalScrollIndicator={false}
            >
                {renderTabs()}
                {tabWidth > 0 && (
                    <Animated.View
                        style={[
                            styles.indicator,
                            {
                                width: tabWidth,
                                backgroundColor: indicatorColor,
                                transform: [{ translateX }],
                            },
                        ]}
                    />
                )}
            </ScrollView>
        </View>
    );
};

const styles = StyleSheet.create({
    wrapper: {
        marginBottom: 12,
    },
    tabRow: {
        flexDirection: 'row',
        borderBottomWidth: 1,
        justifyContent: 'space-around',
        overflow: 'hidden',
        borderRadius: 6,
        position: 'relative',
    },
    tabItem: {
        flex: 1,
        alignItems: 'center',
        paddingVertical: 10,
        minWidth: 80,
    },
    tabText: {
        fontSize: 14,
        fontFamily: 'Poppins',
    },
    indicator: {
        position: 'absolute',
        bottom: 0,
        height: 3,
        borderRadius: 2,
        left: 0,
    },
});

export default Tabs;
