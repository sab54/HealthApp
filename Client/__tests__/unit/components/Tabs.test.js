/**
 * Tabs.test.js
 *
 * What This Test File Covers:
 *
 * 1. Basic Rendering
 *    - Renders all tabs and displays correct labels.
 *
 * 2. Tab Selection Behavior
 *    - Calls onTabSelect when a tab is pressed.
 *    - Does not call onTabSelect for disabled tabs.
 *
 * 3. UI & Theming
 *    - Applies theming to selected, unselected, and disabled states.
 *
 * 4. Animation Styles
 *    - Animates indicator position on tab change.
 *
 * 5. Accessibility
 *    - Tabs are accessible with correct roles and states.
 *
 * 6. Scrollable Tabs
 *    - Renders scrollable container when `scrollable` is true.
 *
 * 7. Snapshots
 *    - Matches snapshot with basic theme.
 */

import React from 'react';
import { render, fireEvent, act } from '@testing-library/react-native';
import Tabs from '@/components/Tabs';
import { Animated } from 'react-native';
import { TouchableOpacity } from 'react-native';

jest.useFakeTimers();
jest.spyOn(Animated, 'spring').mockImplementation(() => ({
  start: (cb) => cb && cb(),
}));

const mockTabs = [
  { key: 'home', label: 'Home' },
  { key: 'search', label: 'Search' },
  { key: 'profile', label: 'Profile' },
];

const themedTabs = [
  { key: 'home', label: 'Home' },
  { key: 'settings', label: 'Settings', disabled: true },
];

const theme = {
  text: '#333',
  primary: '#007AFF',
  surface: '#fafafa',
  border: '#ddd',
  muted: '#aaa',
};

describe('Tabs Component', () => {
  const onTabSelectMock = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders all tab labels', () => {
    const { getByText } = render(
      <Tabs tabs={mockTabs} selectedTab="home" onTabSelect={onTabSelectMock} theme={theme} />
    );

    expect(getByText('Home')).toBeTruthy();
    expect(getByText('Search')).toBeTruthy();
    expect(getByText('Profile')).toBeTruthy();
  });

  it('calls onTabSelect when a tab is pressed', () => {
    const { getByText } = render(
      <Tabs tabs={mockTabs} selectedTab="home" onTabSelect={onTabSelectMock} theme={theme} />
    );

    fireEvent.press(getByText('Profile'));
    expect(onTabSelectMock).toHaveBeenCalledWith('profile');
  });

  it('does not call onTabSelect for disabled tabs', () => {
    const { getByText } = render(
      <Tabs tabs={themedTabs} selectedTab="home" onTabSelect={onTabSelectMock} theme={theme} />
    );

    fireEvent.press(getByText('Settings'));
    expect(onTabSelectMock).not.toHaveBeenCalled();
  });

  it('applies theme styles to selected and disabled tabs', () => {
    const { getByText } = render(
      <Tabs tabs={themedTabs} selectedTab="home" onTabSelect={onTabSelectMock} theme={theme} />
    );

    const homeTab = getByText('Home');
    const settingsTab = getByText('Settings');

    expect(homeTab.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          color: theme.primary,
          fontWeight: '600',
        }),
      ])
    );

    expect(settingsTab.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          color: theme.muted,
        }),
      ])
    );
  });

  it('animates the indicator on selectedTab change', () => {
    const { rerender } = render(
      <Tabs tabs={mockTabs} selectedTab="home" onTabSelect={onTabSelectMock} theme={theme} />
    );

    rerender(
      <Tabs tabs={mockTabs} selectedTab="search" onTabSelect={onTabSelectMock} theme={theme} />
    );

    expect(Animated.spring).toHaveBeenCalledWith(
      expect.any(Animated.Value),
      expect.objectContaining({ toValue: 1, useNativeDriver: false })
    );
  });

  it('assigns correct accessibility props', () => {
    const { UNSAFE_queryAllByType } = render(
      <Tabs tabs={themedTabs} selectedTab="home" onTabSelect={onTabSelectMock} theme={theme} />
    );

    const touchables = UNSAFE_queryAllByType(TouchableOpacity);

    const homeButton = touchables.find(
      (node) => node.props.accessibilityState?.selected === true
    );
    const settingsButton = touchables.find(
      (node) => node.props.accessibilityState?.disabled === true
    );

    expect(homeButton).toBeTruthy();
    expect(homeButton.props.accessibilityRole).toBe('button');
    expect(homeButton.props.accessibilityState).toMatchObject({ selected: true }); // ✅ Only assert present fields

    expect(settingsButton).toBeTruthy();
    expect(settingsButton.props.accessibilityRole).toBe('button');
    expect(settingsButton.props.accessibilityState).toEqual({ selected: false, disabled: true }); // ✅ disabled = true is explicitly present
  });


  it('renders scrollable container when scrollable is true', () => {
    const { UNSAFE_getByType } = render(
      <Tabs tabs={mockTabs} selectedTab="home" onTabSelect={onTabSelectMock} theme={theme} scrollable />
    );

    const scrollView = UNSAFE_getByType(require('react-native').ScrollView);
    expect(scrollView.props.horizontal).toBe(true);
  });

  it('matches snapshot', () => {
    const { toJSON } = render(
      <Tabs tabs={mockTabs} selectedTab="home" onTabSelect={onTabSelectMock} theme={theme} />
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
