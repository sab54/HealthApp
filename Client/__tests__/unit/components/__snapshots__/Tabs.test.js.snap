// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Component matches snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    {
      "marginBottom": 12,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      [
        {
          "borderBottomWidth": 1,
          "borderRadius": 6,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "overflow": "hidden",
          "position": "relative",
        },
        {
          "backgroundColor": "#fafafa",
          "borderColor": "#ddd",
        },
      ]
    }
    horizontal={false}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": true,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "minWidth": 80,
            "opacity": 1,
            "paddingVertical": 10,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "Poppins",
                "fontSize": 14,
              },
              {
                "color": "#007AFF",
                "fontWeight": "600",
              },
            ]
          }
        >
          Home
        </Text>
      </View>
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": false,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "minWidth": 80,
            "opacity": 1,
            "paddingVertical": 10,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "Poppins",
                "fontSize": 14,
              },
              {
                "color": "#333",
                "fontWeight": "500",
              },
            ]
          }
        >
          Search
        </Text>
      </View>
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": false,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "minWidth": 80,
            "opacity": 1,
            "paddingVertical": 10,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "Poppins",
                "fontSize": 14,
              },
              {
                "color": "#333",
                "fontWeight": "500",
              },
            ]
          }
        >
          Profile
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
